File: ros2_ws/src/my_python_pkg/my_python_pkg/hardware_status_publisher.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from my_robot_interfaces.msg import HardwareStatus

class HardwareStatusPublisher(Node):
    def __init__(self):
        super().__init__("hardware_status_publisher")
        self.hardware_status_publisher_ = self.create_publisher(HardwareStatus, "hardware_status", 10)
        self.timer_ = self.create_timer(1.0, self.publish_hardware_status)
        self.get_logger().info("Hardware Status Publisher is running...")

    def publish_hardware_status(self):
        msg = HardwareStatus()
        msg.temperature = 75.0  # Example temperature value
        msg.motor_status = True  # Example motor status
        msg.debug_message = "All systems nominal"  # Example debug info
        self.hardware_status_publisher_.publish(msg)
        self.get_logger().info("Published hardware status")

def main(args=None):
    rclpy.init(args=args)
    node = HardwareStatusPublisher()
    rclpy.spin(node)
    rclpy.shutdown()
 
if __name__ == "__main__":
    main()

File: ros2_ws/src/my_cpp_pkg/src/hardware_status_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/msg/hardware_status.hpp"




class HardwareStatusPublisherNode : public rclcpp::Node {
public:
    HardwareStatusPublisherNode() : Node("hardware_status_publisher")  { 
        hardware_status_publisher_ = this->create_publisher<my_robot_interfaces::msg::HardwareStatus>("hardware_status", 10);
        timer_ = this->create_wall_timer(std::chrono::seconds(1), std::bind(&HardwareStatusPublisherNode::publish_hardware_status, this));
        RCLCPP_INFO(this->get_logger(), "Hardware Status Publisher Node has started publishing hardware status messages.");
    }

private:
    rclcpp::Publisher<my_robot_interfaces::msg::HardwareStatus>::SharedPtr hardware_status_publisher_;
    rclcpp::TimerBase::SharedPtr timer_;

    void publish_hardware_status() {
        auto msg = my_robot_interfaces::msg::HardwareStatus();
        msg.temperature = 75.0; // Example temperature value
        msg.motor_status = true; // Example motor status
        msg.debug_message = "All systems operational"; // Example debug message
        hardware_status_publisher_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "Published hardware status: Temperature: %.2f, Motor Status: %s, Debug Message: %s",
                    msg.temperature, msg.motor_status ? "ON" : "OFF", msg.debug_message.c_str());
    }
};











int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<HardwareStatusPublisherNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}